text/3621-derive-smart-pointer.md

```diff
@@ -0,0 +1,676 @@
+        }
+    }
+}
+```
+In this example, `#[derive(SmartPointer)]` makes it possible to use `Rc<dyn
+MyTrait>`.
+
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+The derive macro will expand into two trait implementations,
+[`core::ops::CoerceUnsized`] to enable unsizing coercions and
+[`core::ops::DispatchFromDyn`] for dynamic dispatch. This expansion will be
+adapted in the future if the underlying mechanisms for unsizing coercions and
+dynamically dispatched receivers changes.
+
+As mentioned in the [rationale][why-only-macro] section, this RFC only proposes
+to stabilize the derive macro. The underlying traits used by its expansion will
+remain unstable for now.
+
+## Input Requirements
+[input-requirements]: #input-requirements
+
+The macro sets the following requirements on its input:
+
+1. The definition must be a struct.
+2. The struct must have at least one type parameter. If multiple type
+   parameters are present, exactly one of them has to be annotated with the
+   `#[pointee]` derive helper attribute.
+3. The struct must not be `#[repr(packed)]` or `#[repr(C)]`.
+4. Other than one-aligned, zero-sized fields, the struct must have exactly one
+   field and that field’s type must be must implement `DispatchFromDyn<F>`
+   where `F` is the type of `T`’s field type.
+
+(Adapted from the docs for [`DispatchFromDyn`].)
+
+Point 1 and 2 are verified syntactically by the derive macro, whereas 3 and 4
+are verified semantically by the compiler when checking the generated
+[`DispatchFromDyn`] implementation as it does today.
+
+## Expansion
+
+The macro will expand to two implementations, one for
+[`core::ops::CoerceUnsized`] and one for [`core::ops::DispatchFromDyn`]. This
+is enough for a type to participe in unsizing coercions and dynamic dispatch.
+
+The derive macro will implement the traits for the type according to the
+following procedure:
+
+- Copy all generic parameters and their bounds from the struct definition into
+  the impl.
+- Add an additional type parameter `U` and give it a `?Sized` bound.
+- Add an additional `Unsize<U>` bound to the `#[pointee]` type parameter.
+- The generic parameter of the traits being implemented will be `Self`, except
+  that the `#[pointee]` type parameter is replaced with `U`.
+
+Given the following example code:
+```rust
+#[derive(SmartPointer)]
+struct MySmartPointer<'a, #[pointee] T: ?Sized, A>{
+    ptr: &'a T
+    phantom: PhantomData<A>
+}
+```
+
+we'll get the following expansion:
+
+```rust
+#[automatically_derived]
+impl<'a, T, A, U> ::core::ops::CoerceUnsized<MySmartPointer<'a, U, A>> for MySmartPointer<'a, T, A>
+where
+    T: ?Sized + ::core::marker::Unsize<U>,
+    U: ?::core::marker::Sized
+
+#[automatically_derived]
+impl<'a, T, A, U> ::core::ops::DispatchFromDyn<MySmartPointer<'a, U, A>> for MySmartPointer<'a, T, A>
+where
+    T: ?Sized + ::core::marker::Unsize<U>,
+    U: ?::core::marker::Sized
+{}
+```
```

Comment:
similar to https://github.com/rust-lang/rfcs/pull/3604#discussion_r1552239797 it is very odd if there isn't also an `impl SmartPointer for MySmartPointer`.

Reply:
This is intentional. I know that there isn't precedent for that in the standard library, but there is lots of precedent in the wider ecosystem. Matching the names of the traits is not so easy, and I discuss why in the [Why not two derive macros?](https://github.com/Darksonn/rfcs/blob/derive-smart-pointer/text/3621-derive-smart-pointer.md#why-not-two-derive-macros) section.

Reply:
I'm not talking about making it `#[derive(CoerceUnsized)]` or `#[derive(DispatchFromDyn)]`, I'm talking about having `#[derive(SmartPointer)]` generate `impl SmartPointer + CoerceUnsized + DispatchFromDyn for S`.

(There has been precedence of a built-in derive generating multiple impls — `#[derive(PartialEq)]` produced `impl PartialEq + StructuralPartialEq for S`.)